lib/src/rules/null_check_on_nullable_type_parameter.dart

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) 2020, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/analysis/features.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+
+import '../analyzer.dart';
+import 'unnecessary_null_checks.dart';
+
+const _desc = r"Don't use null check on a potentially nullable type parameter.";
+
+const _details = r'''
+
+Don't use null check on a potentially nullable type parameter.
+
+Given a generic type parameter `T` which has a nullable bound (e.g. the default
+bound of `Object?`), it is very easy to introduce erroneous null checks when
+working with a variable of type `T?`. Specifically, it is not uncommon to have
+`T? x;` and want to assert that `x` has been set to a valid value of type `T`.
+A common mistake is to do so using `x!`. This is almost always incorrect, since
+if `T` is a nullable type, `x` may validly hold `null` as a value of type `T`.
+
+**BAD:**
+```
+T run<T>(T callback()) {
+  T? result;
+   (() { result = callback(); })();
+  return result!;
+}
+```
+
+**GOOD:**
+```
+T run<T>(T callback()) {
+  T? result;
+   (() { result = callback(); })();
+  return result as T;
+}
+```
+
+''';
+
+class NullCheckOnNullableTypeParameter extends LintRule
+    implements NodeLintRule {
+  NullCheckOnNullableTypeParameter()
+      : super(
+          name: 'null_check_on_nullable_type_parameter',
+          description: _desc,
+          details: _details,
+          maturity: Maturity.experimental,
+          group: Group.style,
+        );
+
+  @override
+  void registerNodeProcessors(NodeLintRegistry registry,
+      [LinterContext context]) {
+    final visitor = _Visitor(this, context);
+    registry.addCompilationUnit(this, visitor);
+    registry.addPostfixExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  bool _isNonNullableEnabled;
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    _isNonNullableEnabled = node.featureSet.isEnabled(Feature.non_nullable);
+  }
+
+  @override
+  void visitPostfixExpression(PostfixExpression node) {
+    if (!_isNonNullableEnabled) return;
+    if (node.operator.type != TokenType.BANG) return;
+
+    final expectedType = getExpectedType(node);
+    if (context.typeSystem.isPotentiallyNullable(node.operand.staticType) &&
```

Comment:
I don't think this quite captures the original request. The original request specified an operand whose type is explicitly nullable (`T?`), where that type is a `TypeParameterType` that is potentially nullable. In particular, I think the test case `m1<T>(T p) => p!;` should not create a lint because in this case the operator isn't being used to remove nullability from the type (which would be more obvious if the method had an explicit return type of `T`).

It might be worthwhile asking on the issue whether that test case ought to produce a lint. It's possible that it was an oversight that the original description doesn't match that case.